Add support for Sorts in numeral notations

JasonGross opened the pull request:
**Kind:** enhancement

- [x] Added / updated test-suite
- [x] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).
- [x] Entry added in CHANGES.md.

proux01 commented on test-suite/output/NumeralNotations.out:
```diff
@@ -184,3 +184,41 @@ let v := 4%Zlike in v : Zlike
+     : ty
+     = 3%kt
+     : ty
+     = 5%kt
```
I would have expected this to be 4%kt (it's should be the result of `Eval cbv in (_ : known_type Set)` if I'm not mistaken)

JasonGross replied:
Hm, good point.  I bet there's an issue with the order of typeclass resolution; I'll fix it.

JasonGross replied:
Fixed

proux01 commented:
Sounds nice! Do you expect this to enable translating R to numeral notation? or what's your intended use?

JasonGross commented:
I don't have a particular use-case in mind; it just seemed like there was no reason to consider sorts as not ground.

Numeral Notations for R (the printing ones at least) still would require support for Ltac printing functions, right?  (It would also require support for types other than just inductives, or else you'd have to wrap the R in a record.)

proux01 commented:
> Numeral Notations for R (the printing ones at least) still would require support for Ltac printing functions, right? (It would also require support for types other than just inductives, or else you'd have to wrap the R in a record.)

If it's just about printing constants in R that are integer, maybe just making IZR a coercion and relying on printing of Z would do the job (@silene ?)

vbgl commented on interp/notation.ml:
```diff
@@ -659,6 +662,10 @@ let rec glob_of_constr token_kind ?loc env sigma c = match Constr.kind c with
+  | Sort Sorts.SProp -> DAst.make ?loc (Glob_term.GSort Glob_term.GSProp)
+  | Sort Sorts.Prop -> DAst.make ?loc (Glob_term.GSort Glob_term.GProp)
+  | Sort Sorts.Set -> DAst.make ?loc (Glob_term.GSort Glob_term.GSet)
+  | Sort (Sorts.Type _) -> DAst.make ?loc (Glob_term.GSort (Glob_term.GType [])) (* FIXME: explain why this list can be empty if its not obvious *)
```
It seems dubious to add code with a FIXME in it. Shall it be addressed before this PR be merged?

JasonGross replied:
@vbgl Sorry for the very late reply.  I left this because it's not clear to me why this list can be empty.  If one of the other devs has an explanation (or claims it's obvious), then I will remove the FIXME.  If no-one can explain it, but it works, then it seems reasonable the leave the FIXME?

vbgl replied:
It seems reasonable to me. Maybe Gaëtan can provide an explanation (@SkySkimmer).

SkySkimmer replied:
This doesn't work anymore as we changed the glob sort representation in https://github.com/coq/coq/pull/8988

JasonGross commented:
I've rebased and updated to the glob sort representation in #8988.  It's not entirely clear to me whether to use `rigid=true` or `rigid=false`, and I didn't see any explanation in #8988 for why we use things like `UNamed [GSet,0]` (in particular, what the `, 0` means).  But it has been languishing for a while, and I think is worth merging.

SkySkimmer commented:
The 0 means +0. I don't know what you do with glob terms so I don't know if the rigidity matters.

JasonGross commented:
As I understand it, parsing deals with `glob_constr`, but all the reduction machinery deals with `constr`.  Since numeral notation parsing needs to invoke reduction, we build a term, pass it as an argument to a registered Gallina function, invoke reduction, and then turn the resulting normal form back into a `glob_constr` so that it is a valid output of prim token parsing.  Ideally, during whatever phase comes after parsing that turns the `glob_constr` back into a `constr`, we will get a `constr` as closely matched as possible to the `constr` output by reduction.

ejgallego commented:
> parsing deals with `glob_constr`

parsing deals with `constr_expr`; `glob_constr` is the structure resulting from the first phase of elaboration.

JasonGross commented:
Windows build is failing with `There is not enough space on the disk.`, which I think is unrelated.

ppedrot commented:
This seems ready, but if nobody merges this before tomorrow, then it'll be postponed. Somebody's assigning?

vbgl commented:
CI failure is spurious: no space left on disk.
